src/jdk.compiler/share/classes/com/sun/tools/javac/util/Iterators.java

```diff
@@ -43,59 +44,45 @@ public static <I, O> Iterator<O> createCompoundIterator(Iterable<I> inputs, Func
         return new CompoundIterator<>(inputs, converter);
     }
 
-    private static class CompoundIterator<I, O> implements Iterator<O> {
+    private static final class CompoundIterator<I, O> implements Iterator<O> {
 
         private final Iterator<I> inputs;
         private final Function<I, Iterator<O>> converter;
-        @SuppressWarnings("unchecked")
-        private Iterator<O> currentIterator = EMPTY;
+        private Iterator<O> currentIterator = Collections.emptyIterator();
 
         public CompoundIterator(Iterable<I> inputs, Function<I, Iterator<O>> converter) {
             this.inputs = inputs.iterator();
             this.converter = converter;
         }
 
+        @Override
         public boolean hasNext() {
             if (currentIterator != null && !currentIterator.hasNext()) {
                 update();
             }
             return currentIterator != null;
         }
 
+        @Override
         public O next() {
-            if (currentIterator == EMPTY && !hasNext()) {
+            if (!hasNext()) {
```

Comment:
I was looking into the history, and the reason for the current code is JDK-8167070 (performance). So, I'd suggest to try how this new code performs. The pre-JDK-8167070 did two calls to nested Iterator's hasNext for each call to hasNext (so for deeper hierarchies, the number of invocations grew very fast), while the code here only does one, so it quite probably may be OK.

I wonder if it is time to write a (unit) test for this class?

Reply:
When I saw a _custom_ empty iterator, my first reaction was to ask myself why it was needed when Collections.emptyIterator had been long available. So yes, I've also seen JDK-8167070.

I think, I was careful enough to preserve the good behavior and remove the bad behavior as well as unneeded code. When developing this PR, I used throwaway unit tests to satisfy myself that the iterator behaves as expected. I'm not sure that we need to integrate that. If we do, then please suggest how it should look like, infrastructure-wise that is.

Reply:
Where in the test tree would you normally put such a unit test?

Reply:
My guess would be to create a directory test/langtools/tools/javac/util/iterators, or some such.

Reply:
I'd use `test/langtools/tools/javac/util`, as I don't think we'll have so many tests for iterators. (Although even the `iterators` subdirectory would not be bad/problematic, subdirectories should be cheap, AFAIK.)